File: python/bucket_sort.py
def bucketSort(array):
  bucket_size = 5
  max_value = max(array)
  min_value = min(array)

  bucket_count = (max_value - min_value) // bucket_size + 1
  buckets = [[] for i in ragne(bucket_count)]

  # insert elements into buckets
  for num in array:
    bucket_idx = (num - min_value) // bucket_size
    buckets[bucket_idx].append(num)

  # sort individual buckets
  for i in range(bucket_count):
    buckets[i].sort()

  # concatenate sorted buckets
  idx = 0
  for bucket in buckets:
    for num in bucket:
      array[idx] = num
      idx += 1


# example usage
array = [170, 45, 75, 90, 802, 24, 2, 66]
bucketSort(array)
print(array)

File: typescript/bucketSort.ts
function bucketSort(arr: number[], bucketSize: number): number[] {
  const maxVal = Math.max(arr);
  const bucketCount = Math.floor(maxVal / bucketSize) + 1;

  const buckets: number[][] = [];
  for(let i = 0; i < bucketCount; i++) {
    buckets.push([]);
  }

  for(let i = 0; i < arr.length; i++) {
    const bucketIdx = Math.floor(arr[i] / bucketSize);
    buckets[bucketIdx].push(arr[i]);
  }
  for(let i = 0; i < bucketCount; i++) {
    buckets[i].sort((a,b) => a - b);
  }
  const sortedArr: number[] = [];
  for(let i = 0; i < bucketCount; i++) {
    sortedArr.push(buckets[i]);
  }
  return sortedArr;
}

// example usage
const numbers = [29, 10, 14, 39, 18, 26, 31, 40];
const sortedNumbers = bucketSort(numbers, 10);
console.log(sortedNumbers);

